extensions/parquet/table/src/main/java/io/deephaven/parquet/table/util/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,273 @@
+package io.deephaven.parquet.table.util;
+
+import io.deephaven.UncheckedDeephavenException;
+import io.deephaven.base.verify.Assert;
+import io.deephaven.parquet.base.util.SeekableChannelsProvider;
+import org.jetbrains.annotations.NotNull;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.SeekableByteChannel;
+import java.security.InvalidParameterException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+
+/**
+ * {@link SeekableByteChannel} class used to fetch objects from AWS S3 buckets using an async client with the ability
+ * to read ahead and cache fragments of the object.
+ */
+public final class S3SeekableByteChannel implements SeekableByteChannel, SeekableChannelsProvider.ContextHolder {
+
+    private static final long CLOSED_SENTINEL = -1;
+
+    /**
+     * Maximum time to wait while fetching fragments from S3.
+     */
+    static final Long CONNECTION_TIMEOUT_MINUTES = 1L;
+    static final TimeUnit CONNECTION_TIMEOUT_UNIT = TimeUnit.MINUTES;
+
+    /**
+     * Context object used to store read-ahead buffers for efficiently reading from S3.
+     */
+    static final class S3ChannelContext implements SeekableChannelsProvider.ChannelContext {
+
+        /**
+         * Used to store information related to a single fragment
+         */
+        static class FragmentContext {
+            /**
+             * The index of the fragment in the object
+             */
+            private final int fragmentIndex;
+
+            /**
+             * The future that will be completed with the fragment's bytes
+             */
+            private final CompletableFuture<ByteBuffer> future;
+
+            private FragmentContext(final int fragmentIndex, final CompletableFuture<ByteBuffer> future) {
+                this.fragmentIndex = fragmentIndex;
+                this.future = future;
+            }
+        }
+
+        private final List<FragmentContext> readAheadBuffers;
```

Comment:
This isn't _just_ read ahead buffers; this is the *cache*. If it were just read ahead, I would expect its size to be `readAheadCount`.